README.md
# Twitch Log Bot
Twitch Log Bot is a bot that will join Twitch channels and store chat messages using Google Cloud Datastore.

The web interface is just a basic interface to allow you to see past messages in a channel to see their chat history.  
It is designed to be simple and only do lookups per-channel (and optionally: per-user).  
If you want something that has more features, take a look at [CBenni's Logviewer](https://github.com/CBenni/logviewer) which is also hosted at [cbenni.com](https://cbenni.com/).

## Setup
- Setup [Google Cloud Datastore](https://console.cloud.google.com/datastore/).
- Setup [node & npm](https://nodejs.org/).
- Install dependencies using `npm install`.
- Copy `config.sample.js` to `config.js` and edit the values (see comments in the config-file).
- Copy `index.sample.yaml` to `index.yaml` and edit each `kind` to what you wish to store it as in Google Cloud Datastore. Anything else should (by default) be left as-is.
    - The easiest way to create these indexes is by installing [Google Cloud SDK](https://cloud.google.com/sdk/), then running `gcloud init` after installing.
    - After that, while being in the directory of this bot, run `gcloud datastore create-indexes index.yaml` and it should create the indexes.
    - If you wanna make sure you don't have unnecessary indexes on GC Datastore, run `gcloud datastore cleanup-indexes index.yaml` in the same directory.
- Copy `channels.sample.json` to `channels.json` and edit it.
    - **It's recommended to at least remove the example channels**.
- Copy `ignore.sample.json` to `ignore.json`.
    - If you are familiar with JSON objects, feel free to manually edit it. Otherwise, I recommend using the [admin commands](COMMANDS.md).
- Run the bot using `node app.js`, or alternatively use [pm2](http://pm2.keymetrics.io/).

## Updating
- Download the [latest release](https://github.com/Decicus/twitch-log-bot/releases/latest) or pull directly from `master` using `git pull`.
- Make sure dependencies are up-to-date by using `npm install`.
- Make sure `config.js` has the same properties as `config.sample.js`.
    - Delete/move config.js elsewhere and make a new `config.js` based off `config.sample.js` **if you feel like you have to**.
- Make sure `index.yaml` has the same values as `index.sample.yaml` (remember to edit the `kind`!) and update indexes using for example `gcloud datastore create-indexes index.yaml`.
- Run the bot using `node app.js`, or alternatively use [pm2](http://pm2.keymetrics.io/).

## Changelog
Changelog can be found here: [CHANGELOG](/CHANGELOG.md).

## Commands
Commands can be found here: [COMMANDS](/COMMANDS.md).

## License
This is licensed under the [MIT License](/LICENSE).


CHANGELOG.md
# Changelog

## Version 0.10.1

- **Subscriptions**: Resubs with no message will no longer include ` - Message: <No Message>` at the end and simply omit it.
- **Subscriptions**: Months should be reported correctly when resubscribing.
    - Resub messages only care about total months subscribed, not "streak months".
- **Subscriptions**: Tiers will now display correctly as `Tier 1/Tier 2/Tier 3` instead of `1000/2000/3000`
    - This only applies to new messages. Older messages already logged will still include `1000/2000/3000`.
    - Twitch Prime subscriptions will still show as `Prime`, unless Twitch changes that in the future.
- **Timeouts / Bans**: Reasons will be omitted from new messages, as timeout/ban reasons are no longer sent via chat.
    - This does not affect timeout lengths, they will still be included as-is.
    - Future bans will now only show up as `* BAN`.
- **Web UI**: The `limit` and `offset` fields will be disabled while retrieving messages, preventing input mid-retrieval.

## Version 0.10.0

- Added `userAgentBlacklist` option in the configuration file.
    - A simple `.includes()` match on the user agent, which returns a `403 Forbidden`.
- Updated npm dependencies.

## Version 0.9.1

- Fix error logs being spammed with exceptions whenever a username that does not exist (banned/deactivated) was looked up.
- Fix messages over 1500 bytes not being logged.
- Updated dependencies.

## Version 0.9.0

- Added support for bans and timeouts, with (timeout) lengths and timeout/ban reason.
- Updated a few project dependencies and the code using said dependencies.

## Version 0.8.3

- Add username to table cell where display name is in certain cases.
    - Specifically where the lowercased version of the display name is different from the username (CJK names).
- Allow people to search for user IDs directly, instead of usernames.
- Fix an issue where channel data (such as IDs) wouldn't be cached when autoconnect wasn't enabled.
    - This would cause errors when trying to fetch messages from a specific channel.

## Version 0.8.2

- Fallback to searching by username instead of user ID when unable to retrieve the ID.
    - Happens in instances where the Twitch API request fails, or the user doesn't "exist" (suspended, deactivated etc).
- Fix "bug" (more of an oversight) where `twitchnotify` messages were being logged as normal messages. These are now ignored.
    - This was introduced in [Version 0.8.0](#version-080) where tmi.js was updated to 1.2.1 and subscription messages were changed.
- Fix `index.sample.yaml` that should have been updated in [Version 0.7.0](#version-070) (I'm sorry).

## Version 0.8.1

- Fix bug where the resub "plan" wasn't logging correctly due to referring to the wrong parameter on the event.

## Version 0.8.0

- Twitch has [done some changes to the subscription system](https://discuss.dev.twitch.tv/t/subscriptions-beta-changes/10023), which allows new subscriptions to share a message, just like resubs.
- Fix bug where quitting the bot wouldn't properly exit the process.

## Version 0.7.0

- Lookups are now done via user IDs instead of names. This will account for users that have changed their names.

## Version 0.6.0

- Added support for logging resub messages. The format is currently:
    - `Resub (X months) - Message: MSG`
    - Where `X` is the number of months and `MSG` is the message the user sent with their resub.
    - If the user didn't specify any message, `<No Message>` will be used as a placeholder.

## Version 0.5.2

- Fix bug/oversight where cheering/bit messages were not being logged.

## Version 0.5.1

- Fix bug where `/me` (aka `ACTION`) messages were not being logged.

## Version 0.5.0

- Adds support for specifying an "offset" in the web interface.
- API for messages now allows query parameters instead of headers.
    - Headers are still allowed to keep it backwards-compatible.
- API now allows messages to be returned in "plaintext", by specifying `plain` as a query parameter or header.

## Version 0.4.2

- Fix error when ignore list for channel did not exist.
- Fix minor error in home view for handling errors.
- Add a `count` field in the `/api/messages` route.

## Version 0.4.1

- Commands `ignore` and `unignore` will now display the correct username that was ignored, instead of the username of the admin that did the command.
    - This is just a minor fix, as the name saved to the ignore list was correct, but the message sent back to the admin was incorrect.

## Version 0.4.0

- Add a per-channel ignore list.
- Added `ignore` and `unignore` commands for use with the ignore list.

## Version 0.3.1

- Adds a 'Refresh' button to the homepage, so that you can retrieve messages without changing from one channel to another, then back.
- Error messages now properly work.

helpers.js
const helpers = {};

/**
 * Formats channel names consistently.
 *
 * @param  {String} channel The channel name to format
 * @return {String}
 */
helpers.fmtChannel = (channel) => {
    return channel.replace('#', '').toLowerCase();
};

/**
 * Returns a new date string of "now" when it's called.
 *
 * @return {String}
 */
helpers.now = () => {
    let date = new Date().toUTCString().split(' ');
    date.shift();
    date.pop();
    return date.join(' ') + ' UTC';
};

/**
 * Formats the timestamp into a UTC date.
 *
 * @param  {Mixed} timestamp The timestamp to format.
 * @return {String}          The formatted date.
 */
helpers.formatDate = (timestamp) => {
    timestamp = parseInt(timestamp);

    let date = new Date(timestamp).toUTCString().split(' ');
    date.shift();
    date.pop();

    return date.join(' ');
};

/**
 * Retrieves either the header or query parameter value from the request.
 *
 * @param  {Object} req  The request object.
 * @param  {String} name The name of the header/query parameter.
 * @return {Mixed}       The value or undefined.
 */
helpers.requestValue = (req, name) => {
    return (typeof req.query[name] !== 'undefined' ? req.query[name] : req.get(name));
};

/**
 * Returns the proper response based on the input value.
 *
 * @param  {Object} res   The express response object.
 * @param  {Mixed} value  The value to send back in the response.
 * @return {Mixed}
 */
helpers.response = (res, value) => {
    if (typeof value !== 'object') {
        res.setHeader('Content-Type', 'text/plain');
        return res.send(value);
    }

    return res.json(value);
};

module.exports = helpers;


config.sample.js
'use strict';

const config = {};

/**
 * See tmi.js docs for options: https://docs.tmijs.org/v1.1.2/Configuration.html
 *
 * @type {Object}
 */
config.tmi = {
    options: {
        debug: true,
    },
    connection: {
        reconnect: true,
        secure: true,
    },
    identity: {
        username: 'decicus',
        // The chat OAuth token can be retrieved using several methods.
        // Here are at least two:
        // - https://decicus.github.io/twitch-chat-token/
        // - https://twitchapps.com/tmi/
        password: 'oauth:KappaHey',
    },
};

/**
 * See Google Cloud options: https://googlecloudplatform.github.io/google-cloud-node/#/docs/google-cloud/0.42.2/google-cloud
 *
 * @type {Object}
 */
config.gcloud = {
    projectId: 'grape-spaceship-123',
    keyFilename: '/path/to/key.json',
};

/**
 * Miscellaneous settings used by the bot.
 *
 * @type {Object}
 */
config.settings = {
    // The path to the JSON-file where channels are stored.
    channels: 'channels.json',
    // The path to the JSON-file where the ignore list is stored.
    ignore: 'ignore.json',
    // The name of the "Kind" on Google Cloud Datastore.
    kind: 'twitch-log-bot',
    // The command prefix for the admin commands.
    prefix: '!',
    // The maximum amount of chat messages the API should allow you to query.
    queryLimit: 200,
    // Array of Twitch usernames that are considered 'admins' and can use whisper commands.
    admins: [],
    // Settings related to when the bot auto-connects to channels on startup.
    autoconnect: {
        // If auto-connecting should be enabled
        enabled: true,
        // The delay between channel join in MILLISECONDS
        // Be careful of not putting this _too low_ due to rate-limits on Twitch's servers.
        delay: 1000,
    },
    // Express-related settings
    express: {
        // Will not run the express/web server if this is set to 'false'
        enabled: false,
        // What port to run the express/web server on
        port: 8000,
    },
    // Twitch API stuff.
    twitch: {
        clientId: '',
    },
    /**
     * Partial matches for blacklisting user agent.
     * E.g. 'Discordbot' will match any user agent that contains 'Discordbot'.
     */
    userAgentBlacklist: ['Discordbot'],
};

module.exports = config;


app.js
'use strict';
const config = require('./config');
const h = require('./helpers');
const pkg = require('./package');
const settings = config.settings;
const datastore = require('@google-cloud/datastore')(config.gcloud);
const express = require('express');
const fs = require('fs');
const request = require('request');
const swig = require('swig');
const tmi = require('tmi.js');

const client = new tmi.client(config.tmi);
const web = express();
const twitchSettings = config.settings.twitch || {};
const baseApi = request.defaults({
    baseUrl: 'https://api.twitch.tv/kraken',
    method: 'GET',
    headers: {
        'Accept': 'application/vnd.twitchtv.v5+json',
        'Authorization': 'OAuth ' + (config.tmi.identity.password || '').replace('oauth:', ''),
        'Client-ID': twitchSettings.clientId || '',
    },
    json: true,
});

/**
 * Channels the bot has joined (and will join on start).
 *
 * @type {Array}
 */
let channels = [];

/**
 * Cache to map usernames to user IDs and vice versa.
 *
 * @type {Object}
 */
let cache = {
    ids: {},
    names: {},
};

/**
 * Get user data based on username.
 *
 * @param  {String}   username
 * @param  {Function} callback
 * @return {Void}
 */
const getUser = (username, callback) => {
    if (typeof callback !== 'function') {
        callback = () => {};
    }

    if (cache.names[username]) {
        callback(cache.names[username]);
        return;
    }

    baseApi({
        'url': '/users?login=' + username,
    }, (err, response, body) => {
        if (err) {
            callback(false);
            return console.error(err);
        }

        if (!body || !body.users || body.users.length === 0) {
            callback(false);
            return console.error(`User ${username} does not exist!`);
        }

        const user = body.users[0];
        const {_id, display_name, name} = user;

        const userToCache = {
            display_name,
            _id,
            name,
        };

        cache.ids[_id] = userToCache;
        cache.names[name] = userToCache;
        console.log(`Loaded ${name} (${_id}) into cache.`);

        callback(userToCache);
    });
};

/**
 * Which users to ignore in channels by the bot.
 *
 * @type {Object}
 */
let ignore = {};

/**
 * Reads a file and parses it as JSON.
 *
 * @return {Mixed}
 */
const readJson = (filename) => {
    try {
        return JSON.parse(fs.readFileSync(filename, 'utf-8'));
    } catch (error) {
        console.log(error);
        return false;
    }
};


/**
 * JSON-encodes the input and writes it to the specified file path.
 *
 * @param  {String} filename
 * @param  {Object|Array} input
 * @return {Void}
 */
const saveJson = (filename, input) => {
    fs.writeFile(filename, JSON.stringify(input, null, 4), (error) => {
        if (error) {
            console.log(error);
        }
    });
};

/**
 * Loads channels from file
 *
 * @return {Void}
 */
const loadChannels = () => {
    let read = readJson(settings.channels || __dirname + '/channels.json');

    if (read === false) {
        return;
    }

    channels = read;
};

// Load channels on startup
loadChannels();

// Cache channel data
channels.forEach((chan) => {
    getUser(chan);
});

/**
 * Saves channels to file.
 *
 * @return {Void}
 */
const saveChannels = () => {
    saveJson(settings.channels || __dirname + '/channels.json', channels);
};

/**
 * Loads the ignore list from file.
 *
 * @return {Void}
 */
const loadIgnore = () => {
    let read = readJson(settings.ignore || __dirname + '/ignore.json');

    if (read === false) {
        return;
    }

    ignore = read;
};

// Load ignore list on startup.
loadIgnore();

/**
 * Saves the ignore list to file.
 *
 * @return {Void}
 */
const saveIgnore = () => {
    saveJson(settings.ignore || __dirname + '/ignore.json', ignore);
};

/**
 * Admin commands controlled by whispers.
 *
 * @type {Object}
 */
const cmds = {};

/**
 * Adds a user to the ignore list for the specified channel, then saves the ignore list.
 */
cmds['ignore'] = (username, user, input) => {
    if (!input[0] || !input[1]) {
        client.whisper(username, 'Both channel and username has to be specified');
        return;
    }

    let channel = h.fmtChannel(input[0]);
    let name = h.fmtChannel(input[1]);

    if (!ignore[channel]) {
        ignore[channel] = [];
    }

    let list = ignore[channel];
    if (list.indexOf(name) >= 0) {
        client.whisper(username, `${name} is already ignored in ${channel}`);
        return;
    }

    list.push(name);
    saveIgnore();
    client.whisper(username, `${name} has been added to the ignore list for ${channel}`);
};

/**
 * Attempts to join the channel if it hasn't already been joined, then saves the channel list.
 */
cmds['join'] = (username, user, input) => {
    if (input[0]) {
        let channel = h.fmtChannel(input[0]);
        if (channels.indexOf(channel) >= 0) {
            client.whisper(username, `${channel} is already joined.`);
        } else {
            getUser(channel, (user) => {
                if (!user) {
                    client.whisper(username, `Error occurred getting data on ${channel}.`);
                    return;
                }

                client.join(channel).then(() => {
                    channels.push(channel);
                    saveChannels();
                    client.whisper(username, `Successfully joined ${channel} and added the channel to the list.`);
                }).catch((err) => {
                    client.whisper(username, `An error occurred joining ${channel}: ${err}`);
                });
            });
        }
    } else {
        client.whisper(username, 'A channel name has to be specified.');
    }
};

/**
 * Leaves the specified channel and removes it from the channel list.
 */
cmds['leave'] = (username, user, input) => {
    if (input[0]) {
        let channel = h.fmtChannel(input[0]);
        let index = channels.indexOf(channel);

        if (index >= 0) {
            client.part(channel).then(() => {
                channels.splice(index, 1);
                saveChannels();
                client.whisper(username, `Successfully left ${channel} and removed the channel from the list.`);
            }).catch((err) => {
                client.whisper(username, `An error occurred leaving ${channel}: ${err}`);
            });
        } else {
            client.whisper(username, `${channel} has not been added to the list yet.`);
        }
    } else {
        client.whisper(username, 'A channel name has to be specified.');
    }
};

/**
 * Alias of cmds.leave().
 */
cmds['part'] = cmds['leave'];

/**
 * Ping pong.
 */
cmds['ping'] = (username) => {
    client.whisper(username, 'PONG');
};

/**
 * Removes a user from the ignore list for the specified channel, then saves the ignore list.
 */
cmds['unignore'] = (username, user, input) => {
    if (!input[0] || !input[1]) {
        client.whisper(username, 'Both channel and username has to be specified');
        return;
    }

    let channel = h.fmtChannel(input[0]);
    let name = h.fmtChannel(input[1]);

    if (!ignore[channel]) {
        ignore[channel] = [];
    }

    let list = ignore[channel];
    let index = list.indexOf(name);
    if (index === -1) {
        client.whisper(username, `${name} isn't ignored in ${channel}`);
        return;
    }

    list.splice(index, 1);
    saveIgnore();
    client.whisper(username, `${name} has been removed from the ignore list for ${channel}`);
};

/**
 * Returns the app version.
 */
cmds['version'] = (username) => {
    client.whisper(username, pkg.version);
};

/**
 * Handles chat messages
 */
const handleMessage = (channel, user, message) => {
    channel = h.fmtChannel(channel);

    /**
     * Ignore legacy "twitchnotify" subscriber messages.
     *
     * These should no longer show up anyways.
     */
    if (user.username === 'twitchnotify') {
        return;
    }

    /**
     * Ignore users in the ignore list.
     */
    if (ignore[channel] && ignore[channel].indexOf(user.username) >= 0) {
        return;
    }

    /**
     * Formats data so we don't log _all_ the junk included in the userstate.
     *
     * @type {Object}
     */
    const data = {
        channel: channel,
        channel_id: user['room-id'],
        username: user.username,
        user_id: user['user-id'],
        user: {
            display_name: (user['display-name'] || user.username),
            type: user['user-type'],
            color: user.color,
            badges: user.badges,
            subscriber: user.subscriber,
            turbo: user.turbo,
        },
        message: message,
        timestamp: user['tmi-sent-ts'],
    };

    const key = datastore.key([settings.kind, user.id]);
    datastore
        .save({
            key: key,
            data: data,
            excludeFromIndexes: ['message'],
        })
        .catch(console.error);
};

/**
 * Defines subscription plans/tiers.
 */
const subTiers = {
    '1000': 'Tier 1',
    '2000': 'Tier 2',
    '3000': 'Tier 3',
};

/**
 * Handles the different subscription events
 */
const handleSubs = (channel, username, m, msg, state, methods) => {
    channel = h.fmtChannel(channel);

    if (ignore[channel] && ignore[channel].indexOf(username) >= 0) {
        return;
    }

    const ts = Date.now().toString();

    getUser(username, (cachedUser) => {
        let user_id = null;
        if (cachedUser) {
            user_id = cachedUser._id;
        }

        /**
         * 'resub' event: This will be the amount of months the user has been subbed for
         * https://github.com/tmijs/docs/blob/gh-pages/_posts/v1.4.2/2019-03-03-Events.md#resub
         *
         * 'subscription' event: This will be an object with the plan/method information
         * https://github.com/tmijs/docs/blob/gh-pages/_posts/v1.4.2/2019-03-03-Events.md#subscription
         */
        const isNumber = typeof m === 'number';

        /**
         * Plan information will be part of `m` if it's a new subscription.
         *
         * Resubs pass another parameter `methods`,
         * which is normally undefined for new subscriptions.
         */
        const {plan, planName} = (isNumber ? methods : m);

        /**
         * Tier 1/2/3 are predefined. If this changes or new tiers show up
         * it should just fall back to whatever Twitch sends via TMI.
         */
        const tier = subTiers[plan] || plan;

        const months = state['msg-param-cumulative-months'] || (isNumber ? m : 0);
        const prefix = isNumber ?
            `Resub (${months} months) - Plan: [${tier}] ${planName}` :
            `Subscription - Plan: [${tier}] ${planName}`;

        /**
         * Check if a resubscription message is included.
         */
        const hasMessage = typeof msg === 'string';
        const data = {
            channel: channel,
            channel_id: cache.names[channel]._id,
            username: username,
            user_id: user_id,
            user: {
                display_name: username,
            },
            message: prefix + (hasMessage ? `- Message: ${msg}` : ''),
            timestamp: ts,
        };

        const key = datastore.key([settings.kind, username + '_' + ts]);
        datastore
            .save({
                key: key,
                data: data,
            })
            .catch(console.error);
    });
};

/**
 * Handles timeouts and bans in a channel.
 *
 * @param {String} channel Channel name where the ban/timeout occurred.
 * @param {String} username Username of user that got banned/timed out.
 * @param {null} reason Historically this contained the ban/timeout reason, but Twitch no longer sends this via TMI.
 * @param {Number|Object|undefined} length Length of time someone got timed out (in seconds).
 */
const handleTimeoutAndBan = (channel, username, timeout, length) => {
    channel = h.fmtChannel(channel);

    const isBan = length === undefined || typeof length === 'object';
    const type = isBan ? 'BAN' : 'TIMEOUT';
    const suffix = isBan ? '' : ` - Length (seconds): ${length}`;

    const ts = Date.now().toString();
    getUser(username, (cachedUser) => {
        let userId = null;
        if (cachedUser) {
            userId = cachedUser._id;
        }

        const data = {
            channel: channel,
            channel_id: cache.names[channel]._id,
            username: username,
            user_id: userId,
            user: {
                display_name: username,
            },
            message: `* ${type + suffix}`,
            timestamp: ts,
        };

        const key = datastore.key([settings.kind, `${channel}_${username}_${ts}`]);
        datastore
            .save({
                key,
                data,
            })
            .catch(console.error);
    });
};

/**
 * Register all chat events with prepared handlers.
 */
client.on('action', handleMessage);
client.on('ban', handleTimeoutAndBan);
client.on('chat', handleMessage);
client.on('cheer', handleMessage);
client.on('resub', handleSubs);
client.on('subscription', handleSubs);
client.on('timeout', handleTimeoutAndBan);

client.on('connected', () => {
    console.log(`[${h.now()}] Successfully connected.`);
    settings.autoconnect = settings.autoconnect || {};

    if (channels.length > 0 && settings.autoconnect.enabled !== false) {
        // concat channels array to get a new array instance.
        let temp = channels.concat();
        let queue = setInterval(() => {
            const chan = temp[0];

            client.join(chan);
            temp.shift();

            if (temp.length === 0) {
                clearInterval(queue);
            }
        }, settings.autoconnect.delay || 1000);
    }
});

client.on('join', (channel, user, self) => {
    if (!self) {
        return;
    }

    console.log(`[${h.now()}] ${user} joined ${channel}`);
});

client.on('whisper', (username, user, message) => {
    username = h.fmtChannel(username);
    /**
     * Handles admin commands through whispers
     */
    if (settings.admins.indexOf(username) >= 0 && message.startsWith(settings.prefix)) {
        let input = message.split(' ');
        // remove prefix from command name
        let command = input[0].slice(settings.prefix.length).toLowerCase();
        if (cmds[command]) {
            // remove the command name from input, because the methods don't need it
            input.shift();
            cmds[command](username, user, input);
        }
    }
});

if (settings.express.enabled) {
    web.engine('html', swig.renderFile);
    web.set('view engine', 'html');
    web.set('views', __dirname + '/views');

    web.get('/', (req, res) => {
        res.render('home');
    });

    web.get('/api/channels', (req, res) => {
        res.send({
            success: true,
            channels: channels,
        });
    });

    /**
     * Blacklisted user agents.
     */
    const blacklistedUserAgents = config.settings.userAgentBlacklist || [];
    web.get('/api/messages', (req, res) => {
        let channel = (h.requestValue(req, 'channel') || '');
        let user = (h.requestValue(req, 'user') || '');
        user = user.trim();
        channel = channel.trim();
        let limit = (parseInt(h.requestValue(req, 'limit')) || 25);
        let offset = (parseInt(h.requestValue(req, 'offset')) || 0);
        let plain = (typeof h.requestValue(req, 'plain') !== 'undefined' ? true : false);
        let useUserId = (typeof h.requestValue(req, 'userid') !== 'undefined' ? true : false);

        let max = (settings.queryLimit || 200);
        if (limit > max) {
            limit = max;
        }

        /**
         * Allow blacklisting of user agents, such as "Discordbot", to prevent their embeds
         * requesting messages.
         */
        const userAgent = req.get('User-Agent');
        for (let i in blacklistedUserAgents) {
            const blacklisted = blacklistedUserAgents[i];
            if (userAgent.includes(blacklisted)) {
                console.log(`Blacklisted user agent: ${userAgent} matching ${blacklisted}`);
                res
                    .status(403)
                    .send({
                        success: false,
                        message: 'Your user agent has been blacklisted.',
                    });

                return;
            }
        }

        if ((!channel || channel.length === 0) && (!user || user.length === 0)) {
            let message = 'No channel or user specified';
            if (plain) {
                h.response(res, message);
                return;
            }

            h.response(res, {
                success: false,
                error: message,
            });
            return;
        }

        let query = datastore.createQuery(settings.kind);

        if (channel && channel.length > 0) {
            query = query.filter('channel_id', '=', cache.names[channel]._id)
                .order('channel_id');
        }

        if (user && user.length > 0) {
            user = user.toLowerCase();
        }

        const userGet = (user && user.length > 0) ? user : channel;
        const handleUser = (cachedUser) => {
            if (!useUserId) {
                if (user && user.length > 0) {
                    /**
                     * User does not "exist", but may just be sitewide suspended/deleted.
                     * Fallback to search by username.
                     *
                     * Eventually I wanna get the user_id from one of the resulting
                     * messages in the query and use that to query more messages by
                     * the user, but as of right now this is just a quickfix.
                     *
                     * This does not account for channels that are banned, which is
                     * something else I'll have to "fix" at a later date.
                     * For now: Don't get suspended.
                     */
                    if (cachedUser) {
                        query = query.filter('user_id', '=', cachedUser._id);
                    } else {
                        query = query.filter('username', '=', user);
                    }
                }
            } else {
                query = query.filter('user_id', '=', cachedUser);
            }

            query = query
                .offset(offset)
                .limit(limit);

            query = query.order('timestamp', {
                descending: true,
            });

            datastore
                .runQuery(query)
                .then((messages) => {
                    messages = messages[0];

                    if (plain) {
                        if (messages.length > 0) {
                            let result = '';
                            for (let index in messages) {
                                let msg = messages[index];

                                result += `[#${msg.channel}][${msg.user.display_name}][${h.formatDate(msg.timestamp)}] - ${msg.message}\r\n`;
                            }

                            h.response(res, result);
                        } else {
                            h.response(res, 'No messages found.');
                        }

                        return;
                    }

                    h.response(res, {
                        success: true,
                        count: messages.length,
                        messages: messages,
                    });
                })
                .catch((err) => {
                    const message = 'Unable to retrieve messages for this user/channel.';
                    console.error(err);
                    res.status(404);

                    if (plain) {
                        h.response(res, message);
                        return;
                    }

                    h.response(res, {
                        success: false,
                        error: message,
                    });
                });
        };

        if (useUserId) {
            handleUser(user);
        } else {
            getUser(userGet, handleUser);
        }
    });

    web.get('/api/status', (req, res) => {
        const state = client.readyState();

        // Set status code based on readyState status.
        const status = (state === 'OPEN' || state === 'CONNECTING') ? 200 : 500;
        res
            .status(status)
            .send({
                success: status === 200,
                state,
            });
    });

    web.get('/api/*', function(req, res) {
        res.send({
            success: false,
            error: '404 not found',
        });
    });

    /**
     * Basic error handling.
     */
    web.use((err, req, res, next) => { // eslint-disable-line no-unused-vars
        if (err.status === 404) {
            res.send('Page not found.');
        }

        res.send('Internal server error');
    });

    let webport = settings.express.port || 8000;
    web.listen(webport, () => {
        console.log(`Web interface listening on port ${webport}`);
    });
}

client.connect();

process.on('SIGINT', () => {
    client.disconnect()
        .then(() => {
            console.log('Successfully disconnected from Twitch chat server.');
            process.exit(0);
        })
        .catch((err) => {
            console.error(`Error occurred disconnecting from Twitch chat server: ${err}`);
            process.exit(1);
        });
});
